File: redis-scripts/leaky_bucket.lua
-- leaky_bucket.lua
-- Atomic leaky bucket with lazy leak computation.
-- KEYS[1] = rate limit key
-- ARGV[1] = capacity, ARGV[2] = leak_rate (per sec), ARGV[3] = now_ms

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local leak_rate = tonumber(ARGV[2])
local now_ms = tonumber(ARGV[3])

local data = redis.call('HMGET', key, 'volume', 'last_leak')
local volume = tonumber(data[1]) or 0
local last_leak = tonumber(data[2]) or now_ms

local elapsed = (now_ms - last_leak) / 1000.0
volume = math.max(0, volume - elapsed * leak_rate)

if volume + 1 <= capacity then
  volume = volume + 1
  redis.call('HMSET', key, 'volume', volume, 'last_leak', now_ms)
  redis.call('PEXPIRE', key, math.ceil((capacity / leak_rate) * 1000) + 1000)
  return {1, 0, math.floor(capacity - volume)}
end

local overflow = volume + 1 - capacity
local retry_ms = math.ceil((overflow / leak_rate) * 1000)
redis.call('HMSET', key, 'volume', volume, 'last_leak', now_ms)
return {0, retry_ms, 0}

File: gateway/internal/algorithms/leaky_bucket.go
package algorithms

import (
	"context"
	"math"
	"sync"
	"time"
)

// LeakyBucket enforces a constant outflow rate using queue-based logic.
//
// Requests enter a queue (bucket). The bucket leaks at a fixed rate.
// If the queue is full, the request is rejected.
//
// Amortized complexity:
//   - Allow: O(1) amortized (lazy leak computation on access)
//   - Space: O(1) per client key
type LeakyBucket struct {
	mu       sync.Mutex
	buckets  map[string]*leakyState
	capacity float64
	leakRate float64
}

type leakyState struct {
	volume   float64
	lastLeak time.Time
}

func NewLeakyBucket(capacity int, leakRate float64) *LeakyBucket {
	return &LeakyBucket{
		buckets:  make(map[string]*leakyState),
		capacity: float64(capacity),
		leakRate: leakRate,
	}
}

func (lb *LeakyBucket) Allow(_ context.Context, key string) (Result, error) {
	lb.mu.Lock()
	defer lb.mu.Unlock()

	now := time.Now()
	st, ok := lb.buckets[key]
	if !ok {
		st = &leakyState{lastLeak: now}
		lb.buckets[key] = st
	}

	elapsed := now.Sub(st.lastLeak).Seconds()
	st.volume = math.Max(0, st.volume-elapsed*lb.leakRate)
	st.lastLeak = now

	if st.volume+1 <= lb.capacity {
		st.volume++
		return Result{
			Allowed:   true,
			Remaining: int(lb.capacity - st.volume),
		}, nil
	}

	overflow := st.volume + 1 - lb.capacity
	retryAfter := time.Duration(overflow/lb.leakRate*float64(time.Second)) + time.Millisecond
	return Result{Allowed: false, RetryAfter: retryAfter, Remaining: 0}, nil
}
